### PR TITLE
fix(module): address shape mismatch for scanvae loss n_samples

### LINKED ISSUE #2429: `SCANVI.loss` fails when `n_samples>1`
Reproducible script:
```
vae = SCANVAE(n_input=100, n_labels=1)
x = torch.randint(0, 100, (10, 100), dtype=torch.float32)
batch = torch.zeros(10, dtype=torch.long)
labels = torch.zeros(10, dtype=torch.long)
tensors = {
    REGISTRY_KEYS.X_KEY: x,
    REGISTRY_KEYS.BATCH_KEY: batch,
    REGISTRY_KEYS.LABELS_KEY: labels,
}

_ = vae.forward(
    tensors,
    inference_kwargs={"n_samples": 2}
)
```

### PR DESCRIPTION
**Pull request by martinkim0**
closes #2429

### CODE REVIEW

File: src/scvi/module/_utils.py [hunk 1]
```diff
@@ -1,45 +1,31 @@
...
+        # return size (batch_size, n_broadcast)
+        return torch.nn.functional.one_hot(labels.squeeze(-1), n_broadcast)
+
+    if o.ndim == 2:
+        batch_size = o.size(0)
+        ys = torch.cat([batch(batch_size, i) for i in range(n_broadcast)])
+        new_o = o.repeat(n_broadcast, 1)
+    elif o.ndim == 3:
+        n_samples = o.size(0)
+        batch_size = o.size(1)
```
Comment (canergen) [inline src/scvi/module/_utils.py:23]: o.size(-2) should work for both.

File: src/scvi/module/_scanvae.py [hunk 2]
```diff
@@ -324,15 +323,26 @@ def loss(
...
-
-        kl_divergence = (kl_divergence_z2.view(self.n_labels, -1).t() * probs).sum(dim=1)
-        kl_divergence += kl(
-            Categorical(probs=probs),
-            Categorical(probs=self.y_prior.repeat(probs.size(0), 1)),
-        )
+        if self.n_labels > 1:
+            reconst_loss += loss_z1_weight + (
+                (loss_z1_unweight).view(self.n_labels, -1).t() * probs
+            ).sum(dim=1)
```
Comment (canergen) [inline src/scvi/module/_scanvae.py:329]: .sum(dim=-1)

File: src/scvi/module/_scanvae.py [hunk 3]
```diff
@@ -324,15 +323,26 @@ def loss(
...
+            kl_divergence = (kl_divergence_z2.view(self.n_labels, -1).t() * probs).sum(dim=1)
```
Comment (canergen) [inline src/scvi/module/_scanvae.py:330]: this line requires a more complex transpose (preferred like torch.transpose(values, -1, -2) ) or requires using torch.einsum.

File: src/scvi/module/_scanvae.py (hunk 2 above)
Comment (canergen) [reply src/scvi/module/_scanvae.py:329]: Remove this check with self.n_labels>1 again.

File: src/scvi/module/_utils.py [hunk 4]
```diff
@@ -1,45 +1,31 @@
 import torch
 
 
-def iterate(obj, func):
```
Comment (canergen) [inline src/scvi/module/_utils.py:4]: Have you removed this function? Is it not in use at other locations?

File: src/scvi/module/_utils.py (hunk 1 above)
Comment (ori-kron-wis) [reply src/scvi/module/_utils.py:23]: not so sure what you meant. o.size(-2) gave me 128 which is like o.size(1) but

Comment (canergen) [reply src/scvi/module/_utils.py:23]: You call here batch_size = o.size(1) and for ndim=2 batch_size = o.size(0). It's nicer to call o.size(-2) and do it outside of the if statement.

File: src/scvi/module/_scanvae.py (hunk 2 above)
Comment (ori-kron-wis) [reply src/scvi/module/_scanvae.py:329]: This comes from the previous solution, which was fine, for the case of n_labels>1 and n_samples==1

File: src/scvi/module/_scanvae.py (hunk 3 above)
Comment (ori-kron-wis) [reply src/scvi/module/_scanvae.py:330]: This comes from the previous solution, which was fine, for the case of n_labels>1 and n_samples==1

File: src/scvi/module/_utils.py (hunk 4 above)
Comment (ori-kron-wis) [reply src/scvi/module/_utils.py:4]: This is from Martin's solution.

File: src/scvi/module/_utils.py (hunk 1 above)
Comment (ori-kron-wis) [reply src/scvi/module/_utils.py:23]: got it.
